netbench/netbench/src/client/thread.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{AddressMap, Client, Connection};
+use crate::{
+    driver::timer::{self, Timer, Timestamp},
+    operation as op, scenario, Checkpoints, Result, Trace,
+};
+use core::{
+    future::Future,
+    task::{Context, Poll},
+};
+use futures::ready;
+
+pub struct Thread<'a, C: Client<'a>> {
+    scenario: &'a scenario::Client,
+    ops: &'a [op::Client],
+    index: usize,
+    op: Option<Op<'a, C>>,
+    timer: Timer,
+}
```

Comment:
non-blocking: curious what the Thread abstraction is for. Are you simulating multi-threaded behavior? Pausing and restating threads?

It also seems like you are using HashSet<u64> as a checkpoint. What exactly is a checkpoint, and how are you using a u64 to control pause and go logic?

Reply:
> Are you simulating multi-threaded behavior? Pausing and restating threads?

Yeah the scenario can have "threads" which are just concurrently executing tasks. For example:
https://github.com/aws/s2n-quic/blob/483667d9e6398146cb1a14c1dfa164647d654e43/netbench/netbench-scenarios/src/request_response.rs#L84-L88

> What exactly is a checkpoint, and how are you using a u64 to control pause and go logic?

It's just a way for one scenario thread to unblock another. For example, let's say we want to start sending on another stream after we've sent 1 MB:

https://github.com/aws/s2n-quic/blob/483667d9e6398146cb1a14c1dfa164647d654e43/netbench/netbench/src/scenario/builder/tests.rs#L53-L82